src/python/pants/backend/java/package/fat_jar.py

```diff
@@ -0,0 +1,241 @@
+from pants.engine.fs import AddPrefix, CreateDigest, Digest, FileContent, MergeDigests, Snapshot
+from pants.engine.internals.selectors import MultiGet
+from pants.engine.process import BashBinary, Process, ProcessResult
+from pants.engine.rules import Get, collect_rules, rule
+from pants.engine.target import (
+    CoarsenedTargets,
+    Targets,
+    TransitiveTargets,
+    TransitiveTargetsRequest,
+)
+from pants.engine.unions import UnionRule
+from pants.jvm.resolve.coursier_fetch import (
+    CoursierLockfileForTargetRequest,
+    CoursierResolvedLockfile,
+    MaterializedClasspath,
+    MaterializedClasspathRequest,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class FatJarFieldSet(PackageFieldSet):
+    required_fields = (
+        JvmMainClassName,
+        JvmRootClassAddress,
+    )
+
+    main_class: JvmMainClassName
+    root_address: JvmRootClassAddress
+    output_path: OutputPathField
+
+
+@rule
+async def package_fat_jar(
+    bash: BashBinary,
+    field_set: FatJarFieldSet,
+) -> BuiltPackage:
+    """
+    Constructs a "fat" JAR file (currently from Java sources only) by
+    1. compiling all Java sources
+    2. building a JAR containing all of those built class files
+    3. creating a fat jar with a broken ZIP index by concatenating all dependency JARs together,
+       followed by the thin JAR we created
+    4. using the unix `zip` utility's repair function to fix the broken fat jar
+    """
+
+    if field_set.root_address.value is None:
+        raise Exception("Needs a `root_address` argument")
+
+    if field_set.main_class.value is None:
+        raise Exception("Needs a `main` argument")
+
+    root_class_address = await Get(
+        Address,
+        AddressInput,
+        AddressInput.parse(field_set.root_address.value, relative_to=field_set.address.spec_path),
+    )
+
+    #
+    # 1. Collect Java source files and compile them all
+    #
+
+    transitive_targets = await Get(
+        TransitiveTargets, TransitiveTargetsRequest([root_class_address])
+    )
+
+    sources_to_compile = await Get(
+        CoarsenedTargets,
+        Addresses(
+            tgt.address for tgt in transitive_targets.closure if tgt.has_field(JavaSourceField)
+        ),
+    )
+
+    javac_request_gets = [
+        Get(FallibleCompiledClassfiles, CompileJavaSourceRequest(tgt)) for tgt in sources_to_compile
+    ]
+    fallible_class_files = await MultiGet(javac_request_gets)
+
+    failed = [i for i in fallible_class_files if i.result == CompileResult.FAILED]
+    if failed:
+        raise Exception(failed[0].stderr)
+
+    # TODO: better handle `i.output is None` case
+    class_files = [i.output for i in fallible_class_files if i.output is not None]
+    compiled_class_files_digest = await Get(Digest, MergeDigests(i.digest for i in class_files))
+    compiled_class_files_snapshot = await Get(Snapshot, Digest, compiled_class_files_digest)
+
+    #
+    # # 2. Produce thin JAR
+    #
```

Comment:
Unrelated to this PR, but: before we go a whole lot further, first party classpaths should be converted into jars anyway. They're much faster to load/operate on than loose classfile directories are. I'll open a ticket.

Reply:
Opened #13036.

Reply:
Yup, sections 1-2 of this function struck me as something that should probably exist elsewhere anyway.